lms/djangoapps/program_enrollments/signals.py

```diff
@@ -0,0 +1,57 @@
+"""
+Signal handlers for program enrollments
+"""
+import logging
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from social_django.models import UserSocialAuth
+from student.models import CourseEnrollmentException
+from third_party_auth.models import SAMLProviderConfig
+
+from lms.djangoapps.program_enrollments.models import ProgramEnrollment
+
+logger = logging.getLogger(__name__)
+
+
+@receiver(post_save, sender=UserSocialAuth)
+def matriculate_learner(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Post-save signal to update any waiting program enrollments with a user,
+    and enroll the user in any waiting course enrollments.
+
+    In most cases this will just short-circuit. Enrollments will only be updated
+    for a SAML provider with a linked organization.
+    """
+    try:
+        user = instance.user
+        provider_slug, external_user_key = instance.uid.split(':')
+        if not SAMLProviderConfig.objects.with_active_flag().filter(is_active=1).get(slug=provider_slug).organization:
+            return
+    except (AttributeError, ValueError, SAMLProviderConfig.DoesNotExist):
+        return
+    except SAMLProviderConfig.MultipleObjectsReturned:
+        logger.warning(
+            u'Unable to activate waiting enrollments for user=%s.'
+            u'  Multiple active SAML configurations found for slug=%s. Expected one.',
+            user.id,
+            provider_slug)
+        return
+
+    incomplete_enrollments = ProgramEnrollment.objects.filter(
+        external_user_key=external_user_key
+    ).prefetch_related('program_course_enrollments')
```

Comment:
I think a `select_related` would work fine here, since it's a FK and not a many-many relationship.  Then you only need to make 1 query with a simple join.